src/experiments/logins/api.js

```diff
@@ -0,0 +1,164 @@
+
+const getLogins = () => {
+  const logins = Services.logins.
+    getAllLogins().
+    filter((l) => !(l.hostname || "").startsWith("chrome://")).
+    map(LoginHelper.loginToVanillaObject);
+  return logins;
+};
+
+const getLogin = (id) => {
+  const login = getLogins().
+    filter(l => l.guid === id)[0];
+  return login || null;
+};
+
+this.logins = class extends ExtensionAPI {
+  getAPI(context) {
+    const EventManager = ExtensionCommon.EventManager;
+    return {
+      experiments: {
+        logins: {
+          // See schema.json for function documentation.
+          getAll() {
+            const logins = getLogins();
+            return logins;
+          },
+          get(id) {
+            const login = getLogin(id);
+            return login;
+          },
+          // TODO: should this be called create(), or should the event name be changed to onAdded?
+          add(loginInfo) {
+            if (getLogin(loginInfo.guid)) {
+              throw new ExtensionError(`Add failed: Login already exists with GUID ${loginInfo.guid}`);
+            }
+            try {
+              const login = LoginHelper.vanillaObjectToLogin(loginInfo);
+              Services.logins.addLogin(login);
+            } catch (ex) {
+              throw new ExtensionError(ex);
+            }
+            const createdLogin = getLogin(loginInfo.guid);
+            return createdLogin;
+          },
+          update(loginInfo) {
+            const login = getLogin(loginInfo.guid);
+            if (!login) {
+              throw new ExtensionError(`Update failed: Login not found with GUID ${loginInfo.guid}`);
+            }
+            const loginAndMetaData = LoginHelper.newPropertyBag(loginInfo);
+            Services.logins.modifyLogin(login, loginAndMetaData);
+            const updatedLogin = getLogin(loginInfo.guid);
+            return updatedLogin;
+          },
+          remove(id) {
+            const login = getLogin(id);
+            if (!login) {
+              throw new ExtensionError(`Remove failed: Login not found with GUID ${id}`);
+            }
+            try {
+              Services.logins.removeLogin(LoginHelper.vanillaObjectToLogin(login));
+            } catch (ex) {
+              throw new ExtensionError(ex);
+            }
+          },
+          onCreated: new EventManager(context, "logins.onCreated", fire => {
+            const callback = (value) => {
+              fire.async(value);
+            };
+            const observer = {
+              observe: (subject, topic, type) => {
+                if (type !== "addLogin") {
+                  return;
+                }
+                subject.QueryInterface(Ci.nsILoginMetaInfo).QueryInterface(Ci.nsILoginInfo);
+                const login = LoginHelper.loginToVanillaObject(subject);
+                callback({ login });
+              },
+            };
+            Services.obs.addObserver(observer, "passwordmgr-storage-changed");
```

Comment:
Standard practice seems to be to wire up a separate observer for each event subscriber, and do the translation from Firefox-internal to API data type inside the callback, like I've done here. Seems odd to me, but I guess we really are only going to have one listener for each event (for now), so, ehh.

Related note: the old extension set up one listener, and then used a case statement to dispatch actions. With this logins API, the addon will wire up the redux store to the LoginManager via a bunch of event listeners:

```js
browser.experiments.logins.onCreated(login => {
  dispatcher.record({
    type: "login_changed",
    mode: "added",
    login: login,
  });
});
// same idea for onUpdated, onRemoved, onAllRemoved
```